include/replace.hpp

```diff
@@ -19,12 +19,76 @@ namespace bit {
 // ========================================================================== //
 
 
-
-// Status: to do
+/*
+ * Case 0: ^10011101^
+ * Case 1: ^10011^101
+ * Case 2: 100^1110^1
+ * Case 3: 10011^101^
+ */
 template <class ForwardIt>
 constexpr void replace(bit_iterator<ForwardIt> first, 
     bit_iterator<ForwardIt> last, bit_value old_value, bit_value new_value) {
-    (first, last, old_value, new_value); 
+
+    if (old_value == new_value) {
+        return;
+    }
+
+    using word_type = typename bit_iterator<ForwardIt>::word_type;
+    constexpr std::size_t num_digits = binary_digits<word_type>::value;
+
+    bit_iterator<ForwardIt> cursor = first;
+    word_type mask;
+    word_type all_ones = _all_ones();
+
+    if (!_is_aligned_lsb(cursor)) {
+        word_type first_word = *cursor.base();
+        if (_in_same_word(cursor, last)) {
+              // Case 2
+              mask = _shift_towards_msb(all_ones, cursor.position());
+              mask &= _shift_towards_lsb(all_ones, num_digits - last.position());
+              if (old_value == bit0) {
+                  *(cursor.base()) |= mask;
+              } else {
+                  *(cursor.base()) &= ~mask;
+              }
+              return;
+        }
+        // Case 1
+        mask = _shift_towards_msb(all_ones, cursor.position());
+
+        if (old_value == bit0) {
+            *(cursor.base()) |= mask;
+        } else {
+            *(cursor.base()) &= ~mask;
+        }
+
+        cursor = bit_iterator(std::next(cursor.base()));
+    }
+
+    ForwardIt word_cursor = cursor.base();
+
+    word_type full_word_value = old_value == bit0 ? _all_ones() : _all_zeros();
+
+    while (word_cursor != last.base()) {
+        // Case 0
+        *word_cursor = full_word_value;
+        ++word_cursor;
+    }
+
```

Comment:
Same comment as in previous PR, I don't see the need to construct a new bit_iterator, but for now its fine.